### PR TITLE
Replace argparse with structopt

### PR DESCRIPTION
**Pull request by robinkrahl**
As discussed in #92, this patch series reworks the argument handling and
replaces argparse with structopt.

### CODE REVIEW

File: nitrocli/src/arg_util.rs [hunk 1]
```diff
@@ -107,14 +107,14 @@ macro_rules! enum_int {
...
-      type Err = ();
+      type Err = &'static str;
 
       fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
         match s {
           $(
             $str => Ok($name::$var),
           )*
-          _ => Err(()),
+          _ => Err("[error]"),
```
Comment (d-e-s-o) [inline nitrocli/src/arg_util.rs]: Can this be a meaningful error? Can we include the string and the available variants? Or is it not user-visible? (looks user-visible to me)

You could use something like this, I suppose:
```
    $all
      .iter()
      .map(::std::convert::AsRef::as_ref)
      .collect::<::std::vec::Vec<_>>()
      .join(",")
```

File: nitrocli/src/arg_util.rs [hunk 2]
```diff
@@ -25,6 +25,58 @@ macro_rules! count {
 }
 
 macro_rules! Command {
+  ( $name:ident, [ $( $var:ident($inner:ident) => ($str:expr, $exec:expr), ) *] ) => {
```
Comment (d-e-s-o) [inline nitrocli/src/arg_util.rs]: Hm, is this a huge code duplication? But I can't seem to find it in the end result, so perhaps it's okay?

File: nitrocli/src/arg_util.rs [hunk 3]
```diff
@@ -76,6 +76,8 @@ macro_rules! Command {
         }
       }
     }
+
+    impl_default!($name => $( $name::$var(::std::default::Default::default()) , )*);
```
Comment (d-e-s-o) [inline nitrocli/src/arg_util.rs]: Seems to me a bit like a misuse of `Default`. Or at least a "dangerous" interpretation of it. But I guess it's only temporary as well?

File: nitrocli/src/args.rs [hunk 4]
```diff
@@ -129,21 +143,271 @@ impl From<DeviceModel> for nitrokey::Model {
...
+/// Interacts with the device’s encrypted volume
+#[derive(Debug, Default, PartialEq, structopt::StructOpt)]
+pub struct EncryptedArgs {
+  #[structopt(subcommand)]
+  subcmd: EncryptedCommand,
+}
+
+/// Closes the encrypted volume on a Nitrokey Storage
+#[derive(Debug, Default, PartialEq, structopt::StructOpt)]
+pub struct EncryptedCloseArgs {}
```
Comment (d-e-s-o) [inline nitrocli/src/args.rs]: Would you mind merging in my patch such that you can remove those empty structs? Alternatively just adding it to the list of commits works, too, if it's too much work.

I think one thing lacking is that we do not currently handle meta information in our macros (e.g., doc attributes). I think you can match on
```rust
#[$docs:meta])* 
```
to accomplish that.

File: nitrocli/src/args.rs [hunk 5]
```diff
@@ -129,21 +143,271 @@ impl From<DeviceModel> for nitrokey::Model {
...
+
+/// Opens the encrypted volume on a Nitrokey Storage
+#[derive(Debug, Default, PartialEq, structopt::StructOpt)]
+pub struct EncryptedOpenArgs {}
+
+/// Interacts with the device’s hidden volume
```
Comment (d-e-s-o) [inline nitrocli/src/args.rs]: Would it be possible to agree on using ' over ’ as apostrophe? I know you always use the latter, but I argue ' is more prevalent in the code base (and I don't even know how to produce ’ to be honest...)

File: nitrocli/src/args.rs [hunk 6]
```diff
@@ -129,21 +143,271 @@ impl From<DeviceModel> for nitrokey::Model {
...
+  /// Shows the name stored on the slot
+  #[structopt(short, long)]
+  name: bool,
+  /// Shows the login stored on the slot
+  #[structopt(short, long)]
+  login: bool,
+  /// Shows the password stored on the slot
+  #[structopt(short, long)]
+  password: bool,
+  /// Prints the sotred data without description
```
Comment (d-e-s-o) [inline nitrocli/src/args.rs]: sorted

**Comment by d-e-s-o** [comment]
I've started going through your changes. Very nice stuff! Thanks for splitting things into reasonable chunks while keeping everything compilable. I look forward to the updated version. By any chance, can you fix the test failures while at it? :-)

File: nitrocli/src/arg_util.rs (hunk 1 above)
Comment (robinkrahl) [reply nitrocli/src/arg_util.rs]: Good catch, I just forgot about it.

File: nitrocli/src/arg_util.rs (hunk 2 above)
Comment (robinkrahl) [reply nitrocli/src/arg_util.rs]: (1) Only the `Display` implementation is duplicate – the other implementations are different (`Command::Variant` vs `Command::Variant(_)`). (2) The old macro variant is deleted after the migration.  In the first implementation, I used internal macros, but that just complicates the transition.

File: nitrocli/src/arg_util.rs (hunk 3 above)
Comment (robinkrahl) [reply nitrocli/src/arg_util.rs]: I think it’s an acceptable implementation, but it’s only temporary anyway.  For structopt, we have to construct an instance of the subcommand enums.  That would be very complicated without `Default`.  The implementation and the macro are removed in the “Remove unused code after structopt transition“ commit.

File: nitrocli/src/args.rs (hunk 5 above)
Comment (robinkrahl) [reply nitrocli/src/args.rs]: I don’t care that much about it, but I can’t promise that I can change my muscle memory. ;)  It is Alt Gr + Shift + N on X11, btw.

**Comment by robinkrahl** [comment]
The failure of the version test is caused by `clap` sending some stuff directly to `stdout`, see https://github.com/clap-rs/clap/issues/1390. I set it to ignore.

**Comment by robinkrahl** [comment]
Thanks for the review! I fixed all issues except the two temporary ones and updated the pull request.

**Comment by robinkrahl** [comment]
Regarding https://github.com/d-e-s-o/nitrocli/commit/c0d8583e0aba69b0628eb9c036f50a86f33d381f – to add support for generating completion scripts, which I intend to do after this is merged, we need to access the `Args` struct from `build.rs`. So it would be easier to just keep it public for the time being.

**Comment by d-e-s-o** [comment]
Merged as https://github.com/d-e-s-o/nitrocli/commit/6c31f18cf40c7662b34ee89b3e6ddf516730611e with some minor changes. Thanks again for your work on this feature!

**Comment by robinkrahl** [comment]
Thanks!

**Pull request closed without merging**